### PR TITLE
Migrate `TransferSchema` to `zod`

### PR DESCRIPTION
**Pull request by iamacook**
## Summary

This migrates the validation of transfers and associated sub-types to `zod`.

## Changes

- Remove `TransferValidator`
- Replace `transferSchema` with `TransferSchema`, consisting of the following replacements:
  - `Erc20TransferSchema`
  - `Erc721TransferSchema`
  - `NativeTokenTransferSchema`
- Infer types from the schemas and propegate them.
- Add appropriate test coverage for the above

### CODE REVIEW

File: src/domain/safe/entities/transfer.entity.ts [hunk 1]
```diff
@@ -1,11 +1,10 @@
...
-  transactionHash: string;
-  transferId: string;
-}
+import { Erc20TransferSchema } from '@/domain/safe/entities/schemas/erc20-transfer.schema';
+import { Erc721TransferSchema } from '@/domain/safe/entities/schemas/erc721-transfer.schema';
+import { NativeTokenTransferSchema } from '@/domain/safe/entities/schemas/native-token-transfer.schema';
+import { TransferSchema } from '@/domain/safe/entities/schemas/transfer.schema';
+import { z } from 'zod';
+
+export type Transfer = z.infer<typeof TransferSchema>;
```
Comment (fmrsabino) [inline src/domain/safe/entities/transfer.entity.ts:7]: nitpick: to move this closer to the schema.

File: src/domain/safe/entities/schemas/erc20-transfer.schema.ts [hunk 2]
```diff
@@ -1,5 +1,21 @@
...
 
+export const Erc20TransferSchema = z.object({
+  type: z.literal('ERC20_TRANSFER'),
+  executionDate: z.coerce.date(),
+  blockNumber: z.number(),
+  transactionHash: HexSchema,
+  to: AddressSchema,
+  from: AddressSchema,
+  value: z.string(),
+  tokenAddress: AddressSchema.nullish().default(null),
```
Comment (fmrsabino) [inline src/domain/safe/entities/schemas/erc20-transfer.schema.ts:14]: This shouldn't be `null`. See https://github.com/safe-global/safe-transaction-service/blob/bb5c5c2f663fdd047cf4e74a9df0fb7da24403f7/safe_transaction_service/history/serializers.py#L845-L853

**Review by fmrsabino — CHANGES_REQUESTED** [review]

File: src/domain/safe/entities/transfer.entity.ts (hunk 1 above)
Comment (iamacook) [reply src/domain/safe/entities/transfer.entity.ts:7]: Moved in ee19ddaff145d1a91fca25d53853eb0c1d8b850c.

File: src/domain/safe/entities/schemas/erc20-transfer.schema.ts (hunk 2 above)
Comment (iamacook) [reply src/domain/safe/entities/schemas/erc20-transfer.schema.ts]: Good catch. Changed in ee19ddaff145d1a91fca25d53853eb0c1d8b850c.

File: src/domain/safe/entities/schemas/erc721-transfer.schema.ts [hunk 3]
```diff
@@ -1,5 +1,21 @@
...
 
+export const Erc721TransferSchema = z.object({
+  type: z.literal('ERC721_TRANSFER'),
+  executionDate: z.coerce.date(),
+  blockNumber: z.number(),
+  transactionHash: HexSchema,
+  to: AddressSchema,
+  from: AddressSchema,
+  tokenId: z.string(),
+  tokenAddress: AddressSchema.nullish().default(null),
```
Comment (fmrsabino) [inline src/domain/safe/entities/schemas/erc721-transfer.schema.ts:14]: The same should apply to the `tokenAddress` of an ERC721 transfer.

**Review by fmrsabino — CHANGES_REQUESTED** [review]

File: src/domain/safe/entities/schemas/erc721-transfer.schema.ts (hunk 3 above)
Comment (iamacook) [reply src/domain/safe/entities/schemas/erc721-transfer.schema.ts]: Fixed in a4fb3f8ccb411e2af223a70d76f90c5e0fce198b.

File: src/routes/transactions/entities/transfers/erc721-transfer.entity.ts [hunk 4]
```diff
@@ -5,8 +5,8 @@ import {
 } from '@/routes/transactions/entities/transfers/transfer.entity';
 
 export class Erc721Transfer extends Transfer {
-  @ApiProperty()
-  tokenAddress: string;
+  @ApiPropertyOptional({ type: String, nullable: true })
+  tokenAddress: `0x${string}` | null;
```
Comment (fmrsabino) [inline src/routes/transactions/entities/transfers/erc721-transfer.entity.ts:9]: Given that it's a required field we should keep it that way on the API level.

**Review by fmrsabino — CHANGES_REQUESTED** [review]

File: src/routes/transactions/entities/transfers/erc721-transfer.entity.ts (hunk 4 above)
Comment (iamacook) [reply src/routes/transactions/entities/transfers/erc721-transfer.entity.ts]: Apologies, missed that! Reverted in 9bdf73d6951ac2ba52fb2643dbf654b3ff98d65d.

**Review by fmrsabino — APPROVED** [review]
🚀

**Review by fmrsabino — APPROVED** [review]
🚀

**Pull request merged**